### ISSUE TITLE
ImageNet DataSource

### ISSUE THREAD

**Issue by nicholas-leonard** [nicholas-leonard/dp#106]
* [x] pull soumiths code
 * [x] script to download and decompress image tars
 * [x] load metadata from devkit to get class id, names and descriptions
 * [x] script to harmonize validation and train datasets
  * [x] get labels from devkit
  * [x] mv all validation images to directories labelled by class
  * [ ] remove all blacklisted images
  * [x] change training set class directory labels from wordnet id to class id.
 * [x] adapt soumiths code to DataSource
 * [x] ImageNet method to measure mean/stdev of pixels
 * [x] TopCrop Feedback which encapsulates [this](https://github.com/facebook/fbcunn/blob/master/examples/imagenet/test.lua#L100)
 * [x] alexnet.lua training script (to test ImageNet)
 * [x] new preprocessing function (ppf) argument for Samplers
 * [x] test using alexnet.lua and th
 * [x] make torchx a dependency
 * [x] documentation (ouch)
  * [x] TopCrop
  * [x] ImageClassSet
  * [x] ImageNet
  * [x] Sampler:async()
  * [x] RandomSampler
 * [x] multi-threading support (mega-ouch)
  * [x] batch = dataset:sub(batch, ...) instead of dataset:sub(batch)
  * [x] will be implemented in ImageClassSet:multithread()
 * [x] fix validation accuracy bug
 * [ ] push new rockspec to rocks
 * [x] final test

**Comment by nicholas-leonard** [comment]
https://github.com/facebook/fbcunn/tree/master/examples/imagenet

**Comment by alexbw** [comment]
+1

**Comment by nicholas-leonard** [comment]
Yeah working on it (#110)

**Comment by alexbw** [comment]
Thanks, I will follow there
